engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAnyNode.java

```diff
@@ -0,0 +1,609 @@
+
+      String selfMessage = null;
+      if (selfInterop.hasExceptionMessage(selfException)) {
+        selfMessage = interop.asString(
+            selfInterop.getExceptionMessage(selfException)
+        );
+      }
+      String otherMessage = null;
+      if (otherInterop.hasExceptionMessage(otherException)) {
+        otherMessage = interop.asString(
+            otherInterop.getExceptionMessage(otherException)
+        );
+      }
+      boolean sameMessages = Objects.equals(selfMessage, otherMessage);
+
+      Object selfCause =
+          selfInterop.hasExceptionCause(selfException) ? selfInterop.getExceptionCause(selfException) : null;
+      Object otherCause =
+          otherInterop.hasExceptionCause(otherException) ? otherInterop.getExceptionCause(otherException) : null;
+      boolean sameCauses = Objects.equals(selfCause, otherCause);
+
+      return sameExceptionTypes && sameMessages && sameCauses;
+    } catch (UnsupportedMessageException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /** Equals for Atoms and AtomConstructors */
+
+  @Specialization
+  boolean equalsAtomConstructors(AtomConstructor selfConstructor, AtomConstructor otherConstructor) {
+    return selfConstructor == otherConstructor;
+  }
+
+  /**
+   * How many {@link EqualsAnyNode} should be created for fields in specialization for atoms.
+   */
+  static final int equalsNodeCountForFields = 10;
+
+  static EqualsAnyNode[] createEqualsNodes(int size) {
+    EqualsAnyNode[] nodes = new EqualsAnyNode[size];
+    Arrays.fill(nodes, EqualsAnyNode.build());
+    return nodes;
+  }
+
+  static InvokeEqualsNode[] createInvokeEqualsNodes(int size) {
+    InvokeEqualsNode[] nodes = new InvokeEqualsNode[size];
+    Arrays.fill(nodes, InvokeEqualsNode.build());
+    return nodes;
+  }
+
+  @Specialization
+  boolean equalsAtoms(Atom self, Atom other,
+      @Cached LoopConditionProfile loopProfile,
+      @Cached(value = "createEqualsNodes(equalsNodeCountForFields)", allowUncached = true) EqualsAnyNode[] fieldEqualsNodes,
+      @Cached(value = "createInvokeEqualsNodes(equalsNodeCountForFields)", allowUncached = true) InvokeEqualsNode[] fieldInvokeEqualsNodes,
+      @Cached InvokeEqualsNode atomInvokeEqualsNode,
+      @Cached ConditionProfile enoughEqualNodesForFieldsProfile,
+      @Cached ConditionProfile fieldsSizeNotEqualProfile,
+      @Cached ConditionProfile constructorsNotEqualProfile) {
+    if (fieldsSizeNotEqualProfile.profile(
+        self.getFields().length != other.getFields().length
+    )) {
+      return false;
+    }
+    if (constructorsNotEqualProfile.profile(
+        self.getConstructor() != other.getConstructor()
+    )) {
+      return false;
+    }
+
+    if (self instanceof Atom selfAtom
+        && other instanceof Atom otherAtom
+        && atomOverridesEquals(selfAtom)) {
+      return atomInvokeEqualsNode.execute(selfAtom, otherAtom);
+    }
```

Comment:
If the atom overrides equals, I think we should allow it to be equal to atoms with a different constructor, right? Then this branch should go _before_ the constructor check - as otherwise the operation will quit early and not give the overridden equality a chance to object.

Reply:
> Wait, aren't self and other already Atom here by the argument types? Why the instanceof checks?

Oops, seems like both me and my IDE have overlooked this.

> If the atom overrides equals, I think we should allow it to be equal to atoms with a different constructor, right? Then this branch should go before the constructor check - as otherwise the operation will quit early and not give the overridden equality a chance to object.

Yes, will move it at the beginning.